### PR TITLE
Fixed #25850 -- Made migrations fail loudly on inconsistent history

### PR DESCRIPTION
**Pull request by Uran198**
I completed fix from #5922 pull request.

### CODE REVIEW

**Comment by timgraham** [comment]
What are the circumstances where the loader check isn't sufficient such that we also need the executor check?

**Comment by Uran198** [comment]
It seems there is no need to make checks in executor. I removed the changes.

File: django/db/migrations/loader.py [hunk 1]
```diff
@@ -6,11 +6,15 @@
 
 from django.apps import apps
 from django.conf import settings
+from django.db import connection as connection_
```
Comment (timgraham) [inline django/db/migrations/loader.py]: We are you using this instead of `self.connection`? Will this work in the multi-database case?

**Comment by timgraham** [comment]
A few cosmetic edits: http://dpaste.com/03YCPA8

File: django/db/migrations/loader.py (hunk 1 above)
Comment (Uran198) [inline django/db/migrations/loader.py]: When `makemigrations` is being run, `MigrationLoader` gets [instantiated](https://github.com/django/django/blob/master/django/core/management/commands/makemigrations.py#L76) with `connection` set to `None`. The `django.db.connection` is a proxy for accessing default `DatabaseWrapper`. So, it will only check the default database in the case of `makemigrations` and chosen with `--database` option, or default if not specified, in case of `migrate`. To check all the databases we can iterate over `django.db.connections`.

Comment (timgraham) [inline django/db/migrations/loader.py]: @shaib, this is your reported ticket -- what's your take on how it should work? Seems to me maybe the check in the executor is more appropriate after all.

Comment (shaib) [inline django/db/migrations/loader.py]: It's a little delicate... neither `makemigrations` nor `migrate` should work at this state, but I think `showmigrations` should work and this breaks it as well.

Comment (shaib) [inline django/db/migrations/loader.py]: The reason `makemigrations` should fail is that the more likely reason for inconsistent history is a change in the migrations code (in particular, dependencies), rather than direct manipulation of the database. If the code of existing migrations is suspicious, we should avoid building more migrations on it.

File: django/db/migrations/loader.py [hunk 2]
```diff
@@ -315,6 +319,32 @@ def _reraise_missing_dependency(migration, missing, exc):
                         # Since we added "key" to the nodes before this implies
                         # "child" is not in there.
                         _reraise_missing_dependency(migration, child, e)
+        # Check that no migrations are applied before their dependencies
+        if self.applied_migrations:
+            applied = set(self.applied_migrations)
```
Comment (MarkusH) [inline django/db/migrations/loader.py]: No need to wrap in `set()`

File: django/db/migrations/loader.py [hunk 3]
```diff
@@ -315,6 +319,32 @@ def _reraise_missing_dependency(migration, missing, exc):
...
+        else:
+            recorder = MigrationRecorder(connection_)
+            applied = set(recorder.applied_migrations())
```
Comment (MarkusH) [inline django/db/migrations/loader.py]: I don't see why you need to fall back to `recorder.applied_migrations()` here. At the beginning of `build_graph()` we ensure `self.applied_migrations` is set.

File: django/db/migrations/loader.py [hunk 4]
```diff
@@ -315,6 +319,32 @@ def _reraise_missing_dependency(migration, missing, exc):
                         # Since we added "key" to the nodes before this implies
                         # "child" is not in there.
                         _reraise_missing_dependency(migration, child, e)
+        # Check that no migrations are applied before their dependencies
```
Comment (MarkusH) [inline django/db/migrations/loader.py]: Can you move the entire block into its own method on the `MigrationLoader`, e.g. `def check_consistent_history(self)` and call that from within `build_graph()`, please.

File: django/db/migrations/loader.py (hunk 1 above)
Comment (MarkusH) [inline django/db/migrations/loader.py]: What @shaib said: `makemigrations` and `migrate` should fail, `showmigrations` should work.

File: django/db/migrations/loader.py (hunk 3 above)
Comment (Uran198) [inline django/db/migrations/loader.py]: In `makemigrations` loader is called with no `connection` and `self.applied_migrations` is [empty](https://github.com/django/django/blob/master/django/db/migrations/loader.py#L172-L173).

File: django/db/migrations/loader.py (hunk 4 above)
Comment (Uran198) [inline django/db/migrations/loader.py]: Maybe, it would be better to call this method from `migrate` and `makemigrations` commands? In this way, it will not be called on `showmigration`.
Also, it could be `check_consistent_history(self, connection)` and in `makemigrations` we can check all database connections.

File: django/db/migrations/loader.py (hunk 1 above)
Comment (Uran198) [inline django/db/migrations/loader.py]: So, we better check all the databases for inconsistent history on `makemigrations`?

Comment (timgraham) [inline django/db/migrations/loader.py]: I guess so.

File: django/db/migrations/loader.py (hunk 4 above)
Comment (timgraham) [inline django/db/migrations/loader.py]: sounds reasonable to me

Comment (shaib) [inline django/db/migrations/loader.py]: Sounds right to me too.

Comment (MarkusH) [inline django/db/migrations/loader.py]: :thumbsup:

File: django/db/migrations/loader.py [hunk 5]
```diff
@@ -316,6 +319,35 @@ def _reraise_missing_dependency(migration, missing, exc):
...
+    def check_consistent_history(self, connection):
+        """
+        Checks that all migrations are applied after their dependencies. Raises
+        InconsistentMigrationHistory exception when migration in applied before
+        it's dependency
+        """
+        recorder = MigrationRecorder(connection)
+        applied = recorder.applied_migrations()
+        for migration in applied:
+            # If we don't know about migration we skip it
```
Comment (MarkusH) [inline django/db/migrations/loader.py]: "... know anything about this migration ..."

File: tests/migrations/test_loader.py [hunk 6]
```diff
@@ -367,3 +369,19 @@ def num_nodes():
...
+    @override_settings(MIGRATION_MODULES={'migrations': 'migrations.test_migrations_inconsistent_history'})
+    def test_check_consistent_history(self):
+        """Loading migrations with inconsistent history should error."""
+        loader = MigrationLoader(connection=None)
+        loader.check_consistent_history(connection)
+        recorder = MigrationRecorder(connection)
+        recorder.record_applied('migrations', '2_auto')
+
+        msg = (
+            "Migration ('migrations', '2_auto') is applied before its "
```
Comment (MarkusH) [inline tests/migrations/test_loader.py]: Nitpicking:
The current migration output uses both `('app_label', 'migration_name')` and `app_label.migration_name` scattered all over the place. I'll open a ticket to use the dot-notation across the board which I find easier to read / write. In the mean time I'd use that format here already.

File: tests/migrations/test_loader.py [hunk 7]
```diff
@@ -367,3 +369,18 @@ def num_nodes():
...
         recorder.flush()
+
+    @override_settings(MIGRATION_MODULES={'migrations': 'migrations.test_migrations_inconsistent_history'})
+    def test_check_consistent_history(self):
+        loader = MigrationLoader(connection=None)
+        loader.check_consistent_history(connection)
+        recorder = MigrationRecorder(connection)
+        recorder.record_applied('migrations', '2_auto')
+
+        msg = (
```
Comment (MarkusH) [inline tests/migrations/test_loader.py]: I think that fits in one line?

**Comment by MarkusH** [comment]
@shaib This looks good to me. Could you sign-off if that's what you had in mind?

File: django/db/migrations/loader.py [hunk 8]
```diff
@@ -316,6 +319,34 @@ def _reraise_missing_dependency(migration, missing, exc):
...
+        applied = recorder.applied_migrations()
+        for migration in applied:
+            # If we don't know anything about this migration we skip it
+            if migration not in self.graph.nodes:
+                continue
+            # Check only parents from the same app
+            unapplied_parents = [
+                x.key for x in self.graph.node_map[migration].parents
+                if x not in applied and x[0] == migration[0]
+            ]
```
Comment (shaib) [inline django/db/migrations/loader.py]: Why are you limiting the check to parents from the same app?

**Comment by shaib** [comment]
Looks mostly right to me. @Uran198 , after you respond to my other comment, it would be helpful if you squashed the commits into one.

File: django/db/migrations/loader.py (hunk 8 above)
Comment (MarkusH) [inline django/db/migrations/loader.py]: Yeah, good question. I was confused when I read it the first time but had an idea why it might be enough. But on second thought that doesn't make sense to me.

Comment (Uran198) [inline django/db/migrations/loader.py]: That was just a quick fix for the issue, and I forgot about it. Without this check I keep getting `Migration admin.0001_initial is applied before its dependency contenttypes.0001_initial` or `Migration admin.0001_initial is applied before its dependency auth.0001_initial`, which is strange. I think the issue may be [here](https://github.com/django/django/blob/master/tests/runtests.py#L150-L156) and the possible solution can be to add `admin: None` to `settings.MIGRATION_MODULES` either in `runtests.py` or in `@override_settings`. I think editing `runtests.py` would be more natural, but I'm not sure if there could be some side effects to this.

Comment (MarkusH) [inline django/db/migrations/loader.py]: Ahh, that's gonna be more interesting. You'd need to look at `self.migrations_module(x[0]) is not None` for whether there are migrations for that particular app or if they are disabled.

Comment (Uran198) [inline django/db/migrations/loader.py]: It probably would not be in a graph if it was disabled. Also, I'm overriding `MIGRATION_MODULES`, so in the loader I don't know which apps are disabled. Even more, it considers all application that are set to `None` in `runtests.py` to be migrated, because of this there is no error, when I add `admin` to it.

Comment (MarkusH) [inline django/db/migrations/loader.py]: I haven't tried it out, but I'm pretty certian this only happens when you use `__first__`: https://github.com/Uran198/django/blob/ticket_25850/django/db/migrations/loader.py#L148 which is silently ignored when adding dependencies.

File: django/db/migrations/loader.py [hunk 9]
```diff
@@ -316,6 +319,33 @@ def _reraise_missing_dependency(migration, missing, exc):
...
+        """
+        recorder = MigrationRecorder(connection)
+        applied = recorder.applied_migrations()
+        for migration in applied:
+            # If we don't know anything about this migration we skip it
+            if migration not in self.graph.nodes:
+                continue
+            unapplied_parents = [
+                x.key for x in self.graph.node_map[migration].parents
+                if x not in applied and settings.MIGRATION_MODULES.get(x[0])
```
Comment (Uran198) [inline django/db/migrations/loader.py]: Maybe, this could be a feasible solution.

File: django/core/management/commands/makemigrations.py [hunk 10]
```diff
@@ -75,6 +76,11 @@ def handle(self, *app_labels, **options):
         # the loader doesn't try to resolve replaced migrations from DB.
         loader = MigrationLoader(None, ignore_no_migrations=True)
 
+        # For every connection check if all migartions are applied after their
```
Comment (timgraham) [inline django/core/management/commands/makemigrations.py]: I think it would be more clear to write "check that there aren't any migrations applied before their dependencies."

File: django/db/migrations/loader.py [hunk 11]
```diff
@@ -316,6 +319,33 @@ def _reraise_missing_dependency(migration, missing, exc):
                         # "child" is not in there.
                         _reraise_missing_dependency(migration, child, e)
 
+    def check_consistent_history(self, connection):
+        """
+        Checks that all migrations are applied after their dependencies. Raises
```
Comment (timgraham) [inline django/db/migrations/loader.py]: use PEP 257 verb style "Check ... Raise .."

File: django/db/migrations/loader.py [hunk 12]
```diff
@@ -316,6 +319,33 @@ def _reraise_missing_dependency(migration, missing, exc):
...
+        InconsistentMigrationHistory exception when migration in applied before
+        it's dependency
```
Comment (timgraham) [inline django/db/migrations/loader.py]: "its"
add period

**Comment by timgraham** [comment]
I guess the tests should go through `migrate` / `makemigrations` (or at least also tested in addition to the existing test). What do you think?

**Comment by Uran198** [comment]
I've added tests to `makemigrations` and `migrate`.

File: django/db/migrations/loader.py (hunk 9 above)
Comment (timgraham) [inline django/db/migrations/loader.py]: I don't think this is correct. `settings.MIGRATION_MODULES` only contains user-defined migration modules -- presumably want the detection to work regardless of whether or not that setting is defined.

Comment (Uran198) [inline django/db/migrations/loader.py]: There is, probably, no need in this or similar check here at all, but in this case, tests will fail yielding inconsistent history with admin module. So, this is a problem only with tests and I don't know what is a proper way to solve this.

File: django/db/migrations/exceptions.py [hunk 13]
```diff
@@ -25,6 +25,13 @@ class CircularDependencyError(Exception):
     pass
 
 
+class InconsistentMigrationHistory(Exception):
+    """
+    Raised when a migration is found to be applied before its dependencies.
```
Comment (MarkusH) [inline django/db/migrations/exceptions.py]: I'd rephrase that to "Raised when an applied migration has some of its dependencies not applied."

File: django/db/migrations/loader.py (hunk 9 above)
Comment (MarkusH) [inline django/db/migrations/loader.py]: I think what you want is this:

```python
unapplied_parents = [
    x.key for x in self.graph.node_map[migraiton].parents
    if x not in applied and (
        x[0] in settings.MIGRATION_MODULES and settings.MIGRATION_MODULES[x[0]] is not None
    )
]
```

where the condition could be changed to

```python
if x not applied and settings.MIGRATION_MODULES.get(x[0], object) is None
```

You probably want to add a comment for that, though, especially for the second code snippet ;)

Comment (MarkusH) [inline django/db/migrations/loader.py]: Furthermore, why do you construct `unapplied_parents` at all? Can't you just loop over `self.graph.node_map[migration].parents` and on raise an exception when the first one has an inconsistent history? That would safe some time in bigger projects.

File: tests/migrations/test_commands.py [hunk 14]
```diff
@@ -462,6 +463,21 @@ def test_migrate_record_squashed(self):
         )
         # No changes were actually applied so there is nothing to rollback
 
+    @override_settings(MIGRATION_MODULES={'migrations': 'migrations.test_migrations_inconsistent_history'})
```
Comment (MarkusH) [inline tests/migrations/test_commands.py]: Do you need that test migration app or can you use another already existing one?

File: django/db/migrations/loader.py (hunk 9 above)
Comment (Uran198) [inline django/db/migrations/loader.py]: @MarkusH Thanks for your proposal, but this condition:
```
x[0] in settings.MIGRATION_MODULES and settings.MIGRATION_MODULES[x[0]] is not None
```
is basically equivalent to `settings.MIGRATION_MODULES.get(x[0]) is not None`. And as @timgraham wrote we want detection to work regardless if that setting is defined.

I'll rewrite it into the loop once this issue will be resolved.

File: tests/migrations/test_commands.py (hunk 14 above)
Comment (Uran198) [inline tests/migrations/test_commands.py]: I can reuse existing one. Thanks! I somehow never thought about it.

File: django/db/migrations/loader.py (hunk 9 above)
Comment (MarkusH) [inline django/db/migrations/loader.py]: Oh, I must have missed the `not` in `is not None`. Yes, `.get()` is equivalent.

**Comment by Uran198** [comment]
From 5ceebfd I created a separate pull request (https://github.com/django/django/pull/6367) in case this will not be merged in.

**Comment by shaib** [comment]
The code looks good to me, please squash the commits and and release notes in the "minor features" section (no normal docs needed IMO).

**Comment by timgraham** [comment]
merged in 02ae5fd31a56ffb42feadb49c1f3870ba0a24869, thanks!
